File: leetcode/双指针/0160_Intersection_of_Two_Linked_Lists/Solution.py
class Solution(object):
    def getIntersectionNode(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """
        if not headA or not headB:
            return None

        p1, p2 = headA, headB

        while p1 != p2:
            p1 = headB if not p1 else p1.next
            p2 = headA if not p2 else p2.next

            # 等同 Java 中, p1 = (p1 == null ? headB : p1.next);

        return p1



class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

File: leetcode/双指针/0160_Intersection_of_Two_Linked_Lists/Solution.java
public class Solution {

    /**
     * 拼接上, 同时遍历
     */
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        if (headA == null || headB == null) {
            return null;
        }

        ListNode p1 = headA, p2 = headB;

        while (p1 != p2) {

            if (p1 == null) {
                p1 = headB;
            } else {
                p1 = p1.next;
            }
  
            if (p2 == null) {
                p2 = headA;
            } else {
                p2 = p2.next;
            }

            // 可以简写成
            // p1 = (p1 == null ? headB : p1.next);
            // p2 = (p2 == null ? headA : p2.next);




            // 注意这么写就错了, 会跳过 B 链表的第一位
            /*
            if (p1 == null) { p1 = headB; }
            if (p2 == null) { p2 = headA; }
            p1 = p1.next;
            p2 = p2.next;
            */
            
        }

        return p1;
    }


    static class ListNode {
        int val;
        ListNode next;

        ListNode(int val) {
            this.val = val;
            this.next = null;
        }
    }
}
